In [ ]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import xgboost as xgb
import seaborn as sb
from sklearn.metrics import auc, accuracy_score, confusion_matrix, mean_squared_error
from sklearn.model_selection import cross_val_score, GridSearchCV, KFold, RandomizedSearchCV, train_test_split
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import f1_score, precision_score, recall_score
from sklearn import preprocessing
from sklearn import linear_model
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
os.chdir('/content/drive/My Drive/Bootcamp ML2020/')
os.listdir('.')
dataset = pd.read_csv('sonar.csv')
dataset.head()

V1      V2      V3      V4  ...     V58     V59     V60  Class
0  0.0200  0.0371  0.0428  0.0207  ...  0.0084  0.0090  0.0032      1
1  0.0453  0.0523  0.0843  0.0689  ...  0.0049  0.0052  0.0044      1
2  0.0262  0.0582  0.1099  0.1083  ...  0.0164  0.0095  0.0078      1
3  0.0100  0.0171  0.0623  0.0205  ...  0.0044  0.0040  0.0117      1
4  0.0762  0.0666  0.0481  0.0394  ...  0.0048  0.0107  0.0094      1

[5 rows x 61 columns]

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      208 non-null    float64
 1   V2      208 non-null    float64
 2   V3      208 non-null    float64
 3   V4      208 non-null    float64
 4   V5      208 non-null    float64
 5   V6      208 non-null    float64
 6   V7      208 non-null    float64
 7   V8      208 non-null    float64
 8   V9      208 non-null    float64
 9   V10     208 non-null    float64
 10  V11     208 non-null    float64
 11  V12     208 non-null    float64
 12  V13     208 non-null    float64
 13  V14     208 non-null    float64
 14  V15     208 non-null    float64
 15  V16     208 non-null    float64
 16  V17     208 non-null    float64
 17  V18     208 non-null    float64
 18  V19     208 non-null    float64
 19  V20     208 non-null    float64
 20  V21     208 non-null    float64
 21  V22     208 non-null    float64
 22  V2

In [ ]:
dataset.describe()

V1          V2          V3  ...         V59         V60       Class
count  208.000000  208.000000  208.000000  ...  208.000000  208.000000  208.000000
mean     0.029164    0.038437    0.043832  ...    0.007941    0.006507    0.466346
std      0.022991    0.032960    0.038428  ...    0.006181    0.005031    0.500070
min      0.001500    0.000600    0.001500  ...    0.000100    0.000600    0.000000
25%      0.013350    0.016450    0.018950  ...    0.003675    0.003100    0.000000
50%      0.022800    0.030800    0.034300  ...    0.006400    0.005300    0.000000
75%      0.035550    0.047950    0.057950  ...    0.010325    0.008525    1.000000
max      0.137100    0.233900    0.305900  ...    0.036400    0.043900    1.000000

[8 rows x 61 columns]

In [ ]:
dataset['Class'].value_counts()

0    111
1     97
Name: Class, dtype: int64

In [ ]:
dataset.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21',
       'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31',
       'V32', 'V33', 'V34', 'V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41',
       'V42', 'V43', 'V44', 'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51',
       'V52', 'V53', 'V54', 'V55', 'V56', 'V57', 'V58', 'V59', 'V60', 'Class'],
      dtype='object')

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      208 non-null    float64
 1   V2      208 non-null    float64
 2   V3      208 non-null    float64
 3   V4      208 non-null    float64
 4   V5      208 non-null    float64
 5   V6      208 non-null    float64
 6   V7      208 non-null    float64
 7   V8      208 non-null    float64
 8   V9      208 non-null    float64
 9   V10     208 non-null    float64
 10  V11     208 non-null    float64
 11  V12     208 non-null    float64
 12  V13     208 non-null    float64
 13  V14     208 non-null    float64
 14  V15     208 non-null    float64
 15  V16     208 non-null    float64
 16  V17     208 non-null    float64
 17  V18     208 non-null    float64
 18  V19     208 non-null    float64
 19  V20     208 non-null    float64
 20  V21     208 non-null    float64
 21  V22     208 non-null    float64
 22  V2

In [ ]:
Y = dataset['Class'] # Target prediksi
x = dataset.drop(['Class'], axis=1) # Data Id tidak diperlukan

In [ ]:
scaler = StandardScaler()
X = scaler.fit_transform(x)
data_x = pd.DataFrame(X, columns=x.columns)

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(data_x, Y, test_size=0.15)

**Logistic Regression**

In [ ]:
logreg = LogisticRegression()
logreg.fit(x_train, y_train)
y_pred = logreg.predict(x_test)
print('Accuracy of logistic regression classifier on train set: {:.2f}'.format(logreg.score(x_train, y_train))) # .score() -> Accuracy
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(x_test, y_test))) # .score() -> Accuracy
print("f1 score         :",f1_score(y_test, y_pred, average='macro'))
print("precision score  :",precision_score(y_test, y_pred, average='macro'))
print("recall score     :",recall_score(y_test, y_pred, average='macro'))

Accuracy of logistic regression classifier on train set: 0.94
Accuracy of logistic regression classifier on test set: 0.66
f1 score         : 0.6559139784946236
precision score  : 0.6607843137254902
recall score     : 0.6626984126984128


In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.61      0.67        18
           1       0.59      0.71      0.65        14

    accuracy                           0.66        32
   macro avg       0.66      0.66      0.66        32
weighted avg       0.67      0.66      0.66        32



**Decision Tree Classifier**


In [ ]:
from sklearn.tree import DecisionTreeClassifier
dectree = DecisionTreeClassifier()
dectree.fit(x_train, y_train)
y_pred = dectree.predict(x_test)
print('Accuracy of Decision Tree Classifier on train set: {:.2f}'.format(dectree.score(x_train, y_train))) # .score() -> Accuracy
print('Accuracy of Decision Tree Classifier on test set: {:.2f}'.format(dectree.score(x_test, y_test))) # .score() -> Accuracy
print("f1 score         :",f1_score(y_test, y_pred, average='macro'))
print("precision score  :",precision_score(y_test, y_pred, average='macro'))
print("recall score     :",recall_score(y_test, y_pred, average='macro'))


Accuracy of Decision Tree Classifier on train set: 1.00
Accuracy of Decision Tree Classifier on test set: 0.72
f1 score         : 0.7184750733137829
precision score  : 0.723529411764706
recall score     : 0.7261904761904762


In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.80      0.67      0.73        18
           1       0.65      0.79      0.71        14

    accuracy                           0.72        32
   macro avg       0.72      0.73      0.72        32
weighted avg       0.73      0.72      0.72        32



**SVM**


In [ ]:
from sklearn.svm import SVC

def kernel_svc(ks):
  classifier_svc = SVC(kernel = ks)
  classifier_svc.fit(x_train, y_train)
  y_pred = classifier_svc.predict(x_test)
  print('Accuracy of SVC ',ks,' classifier on test set: {:.2f}'.format(classifier_svc.score(x_train, y_train))) # .score() -> Accuracy
  print('Accuracy of SVC ',ks,' classifier on test set: {:.2f}'.format(classifier_svc.score(x_test, y_test)))
  print(f1_score(y_test, y_pred, average="macro"))
  print(precision_score(y_test, y_pred, average="macro"))
  print(recall_score(y_test, y_pred, average="macro"))

kernelsvc=['rbf','poly','linear']
for x in kernelsvc:
  print("SVC linear: ", x)
  kernel_svc(x)
  print()

SVC linear:  rbf
Accuracy of SVC  rbf  classifier on test set: 0.98
Accuracy of SVC  rbf  classifier on test set: 0.78
0.7793103448275862
0.7784313725490195
0.7817460317460317

SVC linear:  poly
Accuracy of SVC  poly  classifier on test set: 0.99
Accuracy of SVC  poly  classifier on test set: 0.78
0.7757757757757758
0.7793522267206479
0.7738095238095238

SVC linear:  linear
Accuracy of SVC  linear  classifier on test set: 0.97
Accuracy of SVC  linear  classifier on test set: 0.72
0.7162561576354679
0.7156862745098038
0.7182539682539683



In [ ]:
classifier_linear_svc = SVC()
classifier_linear_svc.fit(x_train, y_train)
y_pred = classifier_linear_svc.predict(x_test)
print('Accuracy of LinearSVC classifier on train set: {:.2f}'.format(classifier_linear_svc.score(x_train, y_train))) # .score() -> Accuracy
print('Accuracy of LinearSVC classifier on test set: {:.2f}'.format(classifier_linear_svc.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of LinearSVC classifier on train set: 0.98
Accuracy of LinearSVC classifier on test set: 0.78
0.7793103448275862
0.7784313725490195
0.7817460317460317


**Random Forest**


In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=300, bootstrap=True, max_features = "sqrt")
rf_model.fit(x_train, y_train)
y_pred = rf_model.predict(x_test)
print('Accuracy of Random Forest on train set: {:.2f}'.format(rf_model.score(x_train, y_train)))
print('Accuracy of Random Forest on test set: {:.2f}'.format(rf_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of Random Forest on train set: 1.00
Accuracy of Random Forest on test set: 0.84
0.8358974358974359
0.8593073593073592
0.8293650793650793


**K-Nearest Neighbors**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.neighbors import KNeighborsRegressor

def k_near(knnr):
  # knear = KNeighborsClassifier(knnr)
  # knear.fit(x_train, y_train)
  # y_pred = knear.predict(x_test)
  # print('Accuracy of KNN on train set: {:.2f}'.format(knear.score(x_train, y_train)))
  # print('Accuracy of KNN on test set: {:.2f}'.format(knear.score(x_test, y_test)))
  # print(f1_score(y_test, y_pred, average="macro"))
  # print(precision_score(y_test, y_pred, average="macro"))
  # print(recall_score(y_test, y_pred, average="macro"))
  
  knn = KNeighborsClassifier(knnr)
  knn.fit(x_train, y_train)
  f1_score(knn.predict(x_train), y_train)
  cross_validation_score = cross_val_score(knn, x_train, y_train, cv=10)
  knn_score = f1_score(knn.predict(x_test), y_test) 
  return knn_score

# k=[1,3,5,7,9,11,13,15,99]
# for x in k:
#   print("K Nearest Neighbor: ", x)
#   k_near(x)
#   print()

k=[1,3,5,7,9,11,13,15,99]
for x in k:
  print("K Nearest Neighbor: ", x)
  print(k_near(x))
  print()

K Nearest Neighbor:  1
0.9285714285714286

K Nearest Neighbor:  3
0.8461538461538461

K Nearest Neighbor:  5
0.7407407407407408

K Nearest Neighbor:  7
0.8275862068965518

K Nearest Neighbor:  9
0.7586206896551724

K Nearest Neighbor:  11
0.689655172413793

K Nearest Neighbor:  13
0.6428571428571429

K Nearest Neighbor:  15
0.689655172413793

K Nearest Neighbor:  99
0.5925925925925927



**XGBoost**

In [ ]:
xgb_model = xgb.XGBClassifier(objective="reg:linear")
xgb_model.fit(x_train, y_train)
y_pred = xgb_model.predict(x_test)
print('Accuracy of XGBoost on train set: {:.2f}'.format(xgb_model.score(x_train, y_train)))
print('Accuracy of XGBoost on test set: {:.2f}'.format(xgb_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

[09:28:22] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
Accuracy of XGBoost on train set: 1.00
Accuracy of XGBoost on test set: 0.75
0.7490196078431373
0.75
0.753968253968254


**Gaussian Process Classifier**

In [ ]:
from sklearn.gaussian_process import GaussianProcessClassifier


gaus_model = GaussianProcessClassifier()
gaus_model.fit(x_train, y_train)

y_pred = gaus_model.predict(x_test)
print('Accuracy of Gauss on train set: {:.2f}'.format(gaus_model.score(x_train, y_train)))
print('Accuracy of Gauss on test set: {:.2f}'.format(gaus_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of Gauss on train set: 1.00
Accuracy of Gauss on test set: 0.91
0.903903903903904
0.9089068825910931
0.9007936507936507


**MLP Classifier**

In [ ]:
from sklearn.neural_network import MLPClassifier
MLP_model = MLPClassifier(hidden_layer_sizes=(3,), activation='logistic',
                       solver='adam', alpha=0.0001,learning_rate='constant', 
                      learning_rate_init=0.3)
MLP_model.fit(x_train, y_train)

y_pred = MLP_model.predict(x_test)
print('Accuracy of MLP on train set: {:.2f}'.format(MLP_model.score(x_train, y_train)))
print('Accuracy of MLP on test set: {:.2f}'.format(MLP_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of MLP on train set: 0.95
Accuracy of MLP on test set: 0.69
0.6875000000000001
0.6984126984126984
0.6984126984126984


**ADA Boost**

In [ ]:
from sklearn.ensemble import AdaBoostClassifier
ada_model = AdaBoostClassifier(n_estimators=100, random_state=0)
ada_model.fit(x_train, y_train)

y_pred = ada_model.predict(x_test)
print('Accuracy of ADA Boost on train set: {:.2f}'.format(ada_model.score(x_train, y_train)))
print('Accuracy of ADA Boost on test set: {:.2f}'.format(ada_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of ADA Boost on train set: 1.00
Accuracy of ADA Boost on test set: 0.81
0.8095238095238095
0.8095238095238095
0.8095238095238095


Dari algoritma algoritma yang saya gunakan, algoritma yang memiliki tingkat akurasi terbaik adalah gauss proccess classifier 91%